### PR TITLE
Keep file extension for temporary files

### LINKED ISSUE #231: Export to excel fails with `GDAL Error (1): File extension should be XLSX` when file exists.
```
INSTALL spatial;
LOAD spatial;

create table tbl (x varchar(200));
insert into tbl values('hallo');
-- straight from the docs
COPY tbl TO 'output.xlsx' WITH (FORMAT GDAL, DRIVER 'xlsx');
```

>  Error: IO Error: GDAL Error (1): File extension should be XLSX

```
duckdb -version
v0.9.2 3c695d7ba9

macOS, 14.1.1 (23B81), m1
```

### PR DESCRIPTION
**Pull request by Maxxen**
When executing a `COPY TO` and the output file already exists, DuckDB by default writes the result to a temporary file which it then replaces the original file with when the copy is done. This temporary file gets the suffix `.tmp`. 

However, when implementing copy functions using external libraries (*cough* GDAL *cough*) they sometimes enforce the file extension as an extra validation step. This then breaks when the temporary file extension is changed.

This PR changes it so that temporary files keep their original extension, but instead get suffixed with a `_tmp` _before_ the extension part. So `foo.csv.tmp` becomes `foo_tmp.csv`. These files are usually not in focus to the user anyway so the exact representation is not that important IMO.

In the process I've also added some additional methods to the `StringUtil` class to help with path manipulation. Im happy to receive feedback on these but I loosely modeled their semantics after the path library in [rust](https://doc.rust-lang.org/std/path/struct.Path.html) (which I think is very well thought-out).

Closes https://github.com/duckdb/duckdb_spatial/issues/231

### CODE REVIEW

**Comment by carlopi** [comment]
I think there might be an untreated case for stuff like:
```
COPY (SELECT 42) TO 'asdasdf' (FORMAT CSV);
```

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! I think this is a good idea - as Carlo mentioned perhaps add some more testing for edge cases?

**Comment by Maxxen** [comment]
Will do!

**Comment by Mytherin** [comment]
Doesn't seem to work as expected still, could you add a test for this?

```sql
D copy (select 42 as x) to 'foo';
D copy (select 42 as x) to 'foo';
Error: IO Error: Cannot open file "/foo_tmp": Read-only file system
```

**Comment by carlopi** [comment]
I would propose to pass down the destination file name to MoveTmpFile, so that instead of two logics (building temporary file from file_path and building file from temporary_file_path), and having to force them to be the inverse of one another, it's just the same logic is applied twice. This guarantees in a more clear way that the file get written to the actual destination, both in the current form and with future modifications.
Unsure how viable is that code-side

Basically something like:
```
	if (op.use_tmp_file) {
                tmp_file = BuildTmpFileName(file);


void PhysicalCopyToFile::MoveTmpFile(ClientContext &context, const string &file_path) {
	auto tmp_file_path = BuildTmpFileName(file_path);
	auto &fs = FileSystem::GetFileSystem(context);
        fs.MoveFile(tmp_file_path, file_path);
}
```

**Comment by Maxxen** [comment]
> I would propose to pass down the destination file name to MoveTmpFile, so that instead of two logics (building temporary file from file_path and building file from temporary_file_path)

That's a good idea but the then we need to save the original file path somewhere, alternatively save the tmp_file path separately in which case we don't need to do any reconstruction at all. I just wanted to avoid having to add anything to the copy operator so as to avoid dealing with serialization.

**Comment by Mytherin** [comment]
Thanks!
